ts/editable/ContentEditable.svelte

```diff
@@ -2,47 +2,71 @@
 Copyright: Ankitects Pty Ltd and contributors
 License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 -->
+<script context="module" lang="ts">
+    export interface ContentEditableAPI {
+        flushLocation(): void;
+    }
+</script>
+
 <script lang="ts">
     import type { Writable } from "svelte/store";
     import { updateAllState } from "../components/WithState.svelte";
+    import type { SelectionLocation } from "../domlib/location";
     import { saveSelection, restoreSelection } from "../domlib/location";
     import { on, preventDefault } from "../lib/events";
     import { caretToEnd } from "../lib/dom";
     import { registerShortcut } from "../lib/shortcuts";
+    import iterateActions from "../sveltelib/iterate-actions";
 
     export let nodes: Writable<DocumentFragment>;
     export let resolve: (editable: HTMLElement) => void;
-    export let mirror: (
-        editable: HTMLElement,
-        params: { store: Writable<DocumentFragment> },
-    ) => void;
 
-    export let inputManager: (editable: HTMLElement) => void;
+    export let mirrors: Array<
+        (editable: HTMLElement, params: { store: Writable<DocumentFragment> }) => void
+    >;
+    export let managers: Array<(editable: HTMLElement) => void>;
```

Comment:
These could do with docstrings

Reply:
Instead of putting docstrings in ContentEditable, I've put doc strings in all libs/interfaces it uses. 

One problem I have with Svelte is, that in `.svelte` files it's hard to quickly overview all the Svelte-y things (actions, listeners, exports) we use for one specific component, because everything is spaced out by all the logic.

That's why I've moved most of the logic into `content-editable.ts` as the "lib" of that very component.

Other candidates for this would be RichTextInput, and PlainTextInput. Which I would do in another PR, if you're okay with this.

Reply:
If the code does not reference things in the .svelte file , then I think putting it in a .ts file is a good idea - it's easier to read/edit that way, and can be tested separately should we wish to.